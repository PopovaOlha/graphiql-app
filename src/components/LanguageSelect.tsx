'use client';

import { usePathname, useRouter } from 'next/navigation';
import { ChangeEvent } from 'react';
import { useTranslation } from 'react-i18next';
import { i18nConfig } from '../../i18nConfig';

const LanguageSelect = () => {
    const { i18n, t } = useTranslation();
    const currentLocale = i18n.language;
    const router = useRouter();
    const currentPathname = usePathname();

    const handleChange = (e: ChangeEvent<HTMLSelectElement>) => {
        const newLocale = e.target.value;

        // set cookie for next-i18n-router
        const days = 30;
        const date = new Date();
        date.setTime(date.getTime() + days * 24 * 60 * 60 * 1000);
        const expires = date.toUTCString();
        document.cookie = `NEXT_LOCALE=${newLocale};expires=${expires};path=/`;

        // redirect to the new locale path
        if (currentLocale === i18nConfig.defaultLocale) {
            router.push('/' + newLocale + currentPathname);
        } else {
            router.push(
                currentPathname.replace(`/${currentLocale}`, `/${newLocale}`)
            );
        }

        router.refresh();
    };

    return (
        <select onChange={handleChange} value={currentLocale}>
            <option value="en">{t('languages.en')}</option>
            <option value="ru">{t('languages.ru')}</option>
        </select>
    );
};

export { LanguageSelect };
