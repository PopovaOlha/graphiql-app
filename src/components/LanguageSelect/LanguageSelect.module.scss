.languageSelect {
    width: 90px;
    margin-right: 40px;
    font-size: 14px;
    color: #fff;
    background-color: #007bff;
    border: none;
    border-radius: 5px;
    padding: 10px 15px;
    appearance: none;
    cursor: pointer;
    transition:
        background-color 0.3s ease,
        box-shadow 0.3s ease;
    box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);

    &:hover {
        background-color: #0056b3;
        box-shadow: 0 6px 12px rgba(0, 86, 179, 0.2);
    }

    &:focus {
        outline: none;
        box-shadow: 0 0 0 3px rgba(0, 123, 255, 0.5);
    }

    option {
        background-color: #fff;
        color: #333;
    }
}
