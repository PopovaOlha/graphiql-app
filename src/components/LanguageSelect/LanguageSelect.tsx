'use client';

import { ChangeEvent } from 'react';
import { useTranslation } from 'react-i18next';
import { FormControl, InputLabel, NativeSelect } from '@mui/material';
import { usePathname, useRouter } from 'next/navigation';

import { i18nConfig } from '../../../i18nConfig';

const LanguageSelect = () => {
    const router = useRouter();
    const currentPathname = usePathname();
    const { i18n, t } = useTranslation();
    const currentLocale = i18n.language;

    const handleChange = (e: ChangeEvent<HTMLSelectElement>) => {
        const newLocale = e.target.value;

        const days = 30;
        const date = new Date();
        date.setTime(date.getTime() + days * 24 * 60 * 60 * 1000);
        const expires = date.toUTCString();
        document.cookie = `NEXT_LOCALE=${newLocale};expires=${expires};path=/`;

        if (currentLocale === i18nConfig.defaultLocale) {
            router.push('/' + newLocale + currentPathname);
        } else {
            router.push(
                currentPathname.replace(`/${currentLocale}`, `/${newLocale}`)
            );
        }

        router.refresh();
    };

    return (
        <FormControl variant="standard" sx={{ m: 1, minWidth: 120 }}>
            <InputLabel variant="standard" htmlFor="language">
                {t('languages.label')}
            </InputLabel>
            <NativeSelect
                defaultValue={currentLocale}
                inputProps={{
                    name: 'language',
                    id: 'language',
                }}
                onChange={handleChange}
            >
                <option value="en">{t('languages.en')}</option>
                <option value="ru">{t('languages.ru')}</option>
            </NativeSelect>
        </FormControl>
    );
};

export { LanguageSelect };
