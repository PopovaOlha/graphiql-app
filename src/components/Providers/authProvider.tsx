import { createContext, Dispatch, SetStateAction, useEffect, useState } from 'react';
import { useIdToken } from 'react-firebase-hooks/auth';

import Preloader from '@/components/Preloader/Preloader';
import { auth } from '@/services/firebase';

export interface IProps {
    children: React.ReactNode;
}

export type TypeSetState<T> = Dispatch<SetStateAction<T>>;

interface AuthContext {
    isAuth: boolean;
    setisAuth?: TypeSetState<AuthContext['isAuth']>;
}

export const AuthContext = createContext<AuthContext>({ isAuth: false });

export default function AuthProvider({ children }: IProps) {
    const [isAuth, setisAuth] = useState(false);
    const [user, loading] = useIdToken(auth);

    useEffect(() => {
        if (user && !loading) {
            setisAuth(true);
        } else if (!user && !loading) {
            setisAuth(false);
        }
    });

    return (
        <AuthContext.Provider value={{ isAuth }}>
            {loading ? <Preloader view={'full'} /> : ''}
            {children}
        </AuthContext.Provider>
    );
}
