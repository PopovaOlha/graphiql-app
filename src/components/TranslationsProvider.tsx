'use client';

import { I18nextProvider } from 'react-i18next';
import initTranslations from '../app/i18n';
import { createInstance, Resource } from 'i18next';
import { FC, ReactNode } from 'react';

export interface TranslationProviderProps {
    children: ReactNode;
    locale: string;
    namespaces: string[];
    resources: Resource;
}

const TranslationsProvider: FC<TranslationProviderProps> = ({
    children,
    locale,
    namespaces,
    resources,
}) => {
    const i18n = createInstance();

    initTranslations(locale, namespaces, i18n, resources);

    return <I18nextProvider i18n={i18n}>{children}</I18nextProvider>;
};

export { TranslationsProvider };
