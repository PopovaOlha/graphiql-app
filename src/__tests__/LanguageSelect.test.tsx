import { fireEvent, render } from '@testing-library/react';
import { Resource } from 'i18next';

import { LanguageSelect } from '@/components/LanguageSelect/LanguageSelect';
import { TranslationsProvider } from '@/components/Providers/TranslationsProvider';

import '@testing-library/jest-dom';

vi.mock('next/navigation', async (importOriginal) => {
    const actual = await importOriginal();
    return {
        ...(actual as object),
        notFound: vi.fn(),
        usePathname: vi.fn(() => ''),
        useRouter: vi.fn(() => ({
            push: vi.fn(),
            refresh: vi.fn(),
        })),
        useParams: vi.fn(() => ({
            method: 'GET',
        })),
        useSearchParams: () => ({
            entries: () =>
                [
                    ['param1', 'value1'],
                    ['param2', 'value2'],
                ][Symbol.iterator](),
        }),
    };
});

vi.mock('react-i18next', async (importOriginal) => {
    const actual = await importOriginal();
    return {
        ...(actual as object),
        I18nextProvider: ({ children }: { children: React.ReactNode }) => (
            <div>{children}</div>
        ),
        useTranslation: () => ({
            t: (key: string) => key,
            i18n: {
                language: 'en',
                changeLanguage: vi.fn(),
            },
        }),
    };
});

const mockResources: Resource = {
    en: {
        mainPage: {
            languages: {
                ru: 'Russian',
                en: 'English',
                label: 'Language',
            },
        },
    },
    ru: {
        mainPage: {
            languages: {
                ru: 'Русский',
                en: 'Английский',
                label: 'Язык',
            },
        },
    },
};

describe('LanguageSelect component', () => {
    it('renders LanguageSelect correctly', () => {
        const { asFragment } = render(
            <TranslationsProvider
                namespaces={['mainPage']}
                locale={'en'}
                resources={mockResources}
            >
                <LanguageSelect />
            </TranslationsProvider>
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it('changes language and sets cookie on selection', () => {
        const { getByRole } = render(
            <TranslationsProvider
                namespaces={['mainPage']}
                locale={'en'}
                resources={mockResources}
            >
                <LanguageSelect />
            </TranslationsProvider>
        );

        const select = getByRole('combobox');
        fireEvent.change(select, { target: { value: 'ru' } });

        expect(document.cookie).toContain('NEXT_LOCALE=ru');
    });
});
