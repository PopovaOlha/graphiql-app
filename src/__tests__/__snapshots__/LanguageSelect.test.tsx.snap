// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LanguageSelect component > renders LanguageSelect correctly 1`] = `
<DocumentFragment>
  <div>
    <div
      class="MuiFormControl-root css-1869usk-MuiFormControl-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-standard MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-standard css-1c2i806-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="true"
        for="language"
      >
        languages.label
      </label>
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl MuiNativeSelect-root css-1eed5fa-MuiInputBase-root-MuiInput-root"
      >
        <select
          aria-invalid="false"
          class="MuiNativeSelect-select MuiNativeSelect-standard MuiInputBase-input MuiInput-input css-1dmqq7i-MuiNativeSelect-select-MuiInputBase-input-MuiInput-input"
          id="language"
          name="language"
        >
          <option
            selected=""
            value="en"
          >
            languages.en
          </option>
          <option
            value="ru"
          >
            languages.ru
          </option>
        </select>
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiNativeSelect-icon MuiNativeSelect-iconStandard css-10bey84-MuiSvgIcon-root-MuiNativeSelect-icon"
          data-testid="ArrowDropDownIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
